apps/src/code-studio/reporting.js

```diff
@@ -244,9 +244,22 @@ reporting.sendReport = function (report) {
   // Post milestone iff the server tells us.
   // Check a second switch if we passed the last level of the script.
   // Keep this logic in sync with ActivitiesController#milestone on the server.
-  if (appOptions.postMilestone ||
-    (appOptions.postFinalMilestone && report.pass && appOptions.level.final_level)) {
+  var postMilestone = appOptions.postMilestone;
+  // Don't do this milestone post if postFailedRunMilestone is enabled,
+  // we did not pass, and this is not the final level (which is frequently
+  // free play and worth saving)
+  if (appOptions.postMilestone && !appOptions.postFailedRunMilestone &&
+    !report.pass && !appOptions.level.final_level) {
+    postMilestone = false;
+  }
+  // if this is the final level and postFinalMilestone is turned on, then
+  // do this post (overrides any earlier decision not to post)
+  if (appOptions.postFinalMilestone && report.pass &&
+    appOptions.level.final_level) {
+    postMilestone = true;
+  }
```

Comment:
Adding a mutable variable into this boolean logic makes this much more difficult to parse/understand. Would it be possible to refactor this down into a single immutable expression?

Here's my attempt:
```
var postMilestone =
  (appOptions.postFinalMilestone && report.pass && appOptions.level.final_level) ||
  !(!appOptions.postFailedRunMilestone && !report.pass) &&
  (appOptions.postMilestone)
```
Or, with inline comments:
```
var postMilestone =
  // Always post if postFinalMilestone is enabled and we pass a final level.
  (appOptions.postFinalMilestone && report.pass && appOptions.level.final_level) ||
  // Don't post if postFailedRunMilestone is disabled and we did not pass.
  !(!appOptions.postFailedRunMilestone && !report.pass) &&
  // Post if postMilestone is enabled.
  (appOptions.postMilestone)
```

Reply:
Even better, extract a helper method, feel free to break the logic up with comments within that method, and verify its behavior with a unit test.